pkg/apis/crd/v1alpha1/types.go

```diff
@@ -354,3 +355,123 @@ type BGPPeer struct {
+type TransportHeader struct {
+	UDP *UDPHeader `json:"udp,omitempty"`
+	TCP *TCPHeader `json:"tcp,omitempty"`
+}
+
+// UDPHeader describes spec of a UDP header.
+type UDPHeader struct {
+	// SrcPort is the source port.
+	SrcPort *int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort *int32 `json:"dstPort,omitempty"`
+}
+
+// TCPHeader describes spec of a TCP header.
+type TCPHeader struct {
+	// SrcPort is the source port.
+	SrcPort *int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort *int32 `json:"dstPort,omitempty"`
+	// Flags are flags in the header.
+	Flags *int32 `json:"flags,omitempty"`
+}
+
+// Packet includes header info.
+type Packet struct {
+	// IPFamily is the filter's IP family. Default to `IPv4`.
+	IPFamily v1.IPFamily `json:"ipFamily"`
+	// Protocol represents the transport protocol. default to ICMP(1). Other
+	// possible choices are: TCP(6), UDP(17).
+	Protocol        *intstr.IntOrString `json:"protocol,omitempty"`
+	TransportHeader TransportHeader     `json:"transportHeader"`
+}
+
+// PacketCaptureFirstNConfig contains the config for the FirstN type capture. The only supported parameter is
+// `Number` at the moment, meaning capturing the first specified number of packets in a flow.
+type PacketCaptureFirstNConfig struct {
+	Number int32 `json:"number"`
+}
+
+const DefaultPacketCaptureTimeout uint16 = 60
+
+type PacketCapturePhase string
+
+const (
+	PacketCaptureRunning   PacketCapturePhase = "Running"
+	PacketCaptureSucceeded PacketCapturePhase = "Succeeded"
+	PacketCaptureFailed    PacketCapturePhase = "Failed"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type PacketCaptureList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []PacketCapture `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type PacketCapture struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              PacketCaptureSpec   `json:"spec"`
+	Status            PacketCaptureStatus `json:"status"`
+}
+
+type CaptureConfig struct {
+	FirstN *PacketCaptureFirstNConfig `json:"firstN,omitempty"`
+}
+
+type PacketCaptureSpec struct {
+	Timeout       *uint16       `json:"timeout,omitempty"`
+	CaptureConfig CaptureConfig `json:"captureConfig"`
+	Source        Source        `json:"source"`
+	Destination   Destination   `json:"destination"`
+	Packet        *Packet       `json:"packet,omitempty"`
+	// FileServer specifies the sftp url config for the fileServer. Captured packets will be uploaded to this server.
+	FileServer BundleFileServer `json:"fileServer"`
+}
+
+type PacketCaptureStatus struct {
+	Phase PacketCapturePhase `json:"phase"`
+	// Reason records the failure reason when the capture fails.
+	Reason string `json:"reason"`
```

Comment:
Not sure a single Phase and Reason are enough for all conditions that it can occur when capturing packets on two Nodes and uploading to a server, maybe it will need a list conditions to represent each condition and respective reason and timestamp. But hard to say what are needed until it's implemented, we can defer to the implementation patch.

Reply:
you mean conditions like `Traceflow`? Currently i didn't implement this in PacketCapture, not sure if this is needed.

Reply:
Phase is a deprecated pattern of K8s API, it recommends to use conditions which is more flexible to express state of each condition. For example, there could be a condition for capture and another condition for upload. See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

I don't have strong opinion if the phase and a single reason can cover all possible results.

Reply:
updated. use `conditions` instead of `Phase/Reason`